### PR TITLE
Optimize fiopen

### PR DESCRIPTION
**Pull request by AtariDreams**
Also the if statements are now consistent with returning nullptr now.

### CODE REVIEW

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:62]: If `mode == 0`, this function previously returned `nullptr` here but now continues processing.

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:78]: `n` on the line above used to denote the index determined by the loop; I'm not sure what (if anything) it now names.

File: stl/src/fiopen.cpp
Comment (AtariDreams) [reply stl/src/fiopen.cpp:63]: in that case, let's save time by checking

Comment (AtariDreams) [reply stl/src/fiopen.cpp:63]: mode being 0 should be the first thing we check, not the last in that case, since there is no avoiding this check anyway

File: stl/src/fiopen.cpp
Comment (AtariDreams) [reply stl/src/fiopen.cpp:82]: It still does now.

**Comment by AtariDreams** [comment]
All issues addressed!

File: stl/src/fiopen.cpp
Comment (AlexGuteniev) [inline stl/src/fiopen.cpp:62]:
```suggestion
    if (mode == 0) {
        return nullptr; // no valid mode
    }
```

**Comment by AtariDreams** [comment]
Done!

File: stl/src/fiopen.cpp
Comment (AlexGuteniev) [inline stl/src/fiopen.cpp:65]: Not sure what is better:
 * Have apparent missing initialization (keep it as is)?
 * Have odd `for` loop (move `= 0` to here from for init, keep for init empty)?

Comment (AtariDreams) [reply stl/src/fiopen.cpp]: Done!

**Comment by AtariDreams** [comment]
Addressed!

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:69]: This doesn't strike me as an improvement: we started with two branches poor loop iteration, and still have two branches per loop iteration, and the trivially-optimized-away branch after the loop has become a very hard to optimize away branch before the loop. I think we could improve a bit by abandoning standard loop structure:
```c++
int n = 0;
for (;; ++n) {
  if (valid[n] == 0) {
    return nullptr;
  }
  if (valid[n] == mode) {
    break;
  }
}
```
which coalesces the before-or-after the loop check into the per-iteration checks (which is the form a good optimizer likely transforms the original code into anyway).

Comment (AlexGuteniev) [reply stl/src/fiopen.cpp:69]: Is the goal performance? It's odd to make instruction level optimizations for file i/o (which is ultimately a syscall), and it is odd to favor negative case in such optimization

Comment (AtariDreams) [reply stl/src/fiopen.cpp:69]: Negative case yes, but at the same time, this check is always done anyway. If it wasn’t, then you’d be right. 

If checks are inevitable, put them first so in those cases we can save time

**Comment by AtariDreams** [comment]
Addressed!

File: stl/src/fiopen.cpp
Comment (AlexGuteniev) [inline stl/src/fiopen.cpp:50]: I think this part is unnecessary. It is odd to optimize _performance_ of negative case by making positive case less optimal. And I don't think a function that is supposed to open a file needs fine instruction-level optimization.

I would drop this part of change, and keep the rest as an optimization of _readability_.

**Review by AlexGuteniev — CHANGES_REQUESTED** [review]

File: stl/src/fiopen.cpp
Comment (AtariDreams) [reply stl/src/fiopen.cpp:50]: Positive case is actually NOT less optimal.

Negative case ALWAYS has to be checked first because otherwise it will cause a bug if mode == 0

Comment (AtariDreams) [reply stl/src/fiopen.cpp:50]: I took this into serious consideration with this patch. Positive case is the same. Optimizing for the negative case without affecting the positive is a good thing

Comment (AtariDreams) [reply stl/src/fiopen.cpp:50]: The original code also checks for the negative first.

**Comment by AtariDreams** [comment]
Positive case is actually NOT less optimal.

Negative case ALWAYS has to be checked first because otherwise it will cause a bug if mode == 0

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:72]: If we reorder these two conditions - check for `0` and `return` before checking for equality and ending the loop - then we can drop the separate check on lines 47-50. The performance of correct calls is unaffected, performance of erroneous calls potentially slowed (which we don't care about), and there's a small reduction in code size.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/src/fiopen.cpp
Comment (StephanTLavavej) [inline stl/src/fiopen.cpp:70]: Style: We conventionally separate "non-chained" `if` statements with a newline. This makes it easier to see when we have independent tests, versus an `if` chained to `else if`.

File: stl/src/fiopen.cpp
Comment (AtariDreams) [reply stl/src/fiopen.cpp:72]: Erroneous? Well, we don't know if it is erroneous, but if it is, I can fix it back

**Comment by AtariDreams** [comment]
Addressed!

**Comment by AtariDreams** [comment]
Any updates?

**Comment by AlexGuteniev** [comment]
> Any updates?

Seing other issues in maintainers plan helps having some patience https://github.com/microsoft/STL/projects/6

**Comment by AtariDreams** [comment]
All issues addressed! I had to fix conflicts, but it works now!

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:62]: If a user passes a value in `mode` that does not appear in `valid`, this loop runs off the end of `valid`.

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:63]:
```suggestion
        if (n == size(valid)) {
```
(Don't forget to `#include <iterator>`.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/src/fiopen.cpp
Comment (CaseyCarter) [inline stl/src/fiopen.cpp:66]: We need to terminate the loop _before_ examining `valid[size(valid)]`, which is past-the-end of the array:
```suggestion
    while (valid[n] != mode) {
        if (++n == _STD size(valid)) {
            return nullptr; // no valid mode
        }
    }
```

**Review by CaseyCarter — APPROVED** [review]

File: stl/src/fiopen.cpp
Comment (StephanTLavavej) [inline stl/src/fiopen.cpp:63]: x86 builds are failing:

```
C:\a\1\s\stl\src\fiopen.cpp(63,17): warning C4389: '==': signed/unsigned mismatch
        if (++n == _STD size(valid)) {
                ^
```

Comment (CaseyCarter) [inline stl/src/fiopen.cpp:63]:
```suggestion
        if (++n == static_cast<int>(_STD size(valid))) {
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - looks good to me, I believe this is ready to merge. However, due to the [VC Redist Lockdown](https://github.com/microsoft/STL/wiki/Changelog#vc-redist-lockdown), we'll need to hold off on merging this for a hopefully short time, until internal testing with the "locked redist" is added (currently being investigated by @MahmoudGSaleh). Once that's in place, we can merge this change as it affects `msvcp140.dll` but does not require synchronized changes to the headers.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for simplifying this code and eliminating the weird zero-termination of that array! :tada: :rocket: :smiley_cat:

**Pull request merged**
